### ISSUE TITLE
uploading large file 3GB or more as readablestream consume huge amount of memory.

### ISSUE THREAD

**Issue by Guling85** [tus/tus-js-client#363]
Im trying to upload a large file around 3GB with TUS client. It seems to me that TUS client doesnt use readablestream as intended because the memory usage is huge. The memory usage of chrome is like 2.5GB for 3 GB file.

here is my code.

```
 const file = $event.target.files[0];

      const upload = new Upload(file.stream().getReader(), {
        endpoint: "https://localhost:8443/upload/file",
        retryDelays: [0, 3000, 6000, 12000, 24000],
        chunkSize: 6553600,
        uploadLengthDeferred: true,
        metadata: {
          uuid: result[0].uploadId,
          order: chunkOrder.toString()
        },
        //chunkSize: 1048576,
        onBeforeRequest: (req) => {
          const xhr = req.getUnderlyingObject();
          xhr.withCredentials = false;
        },
        onError: async (error) => {
          console.log(error);

        },
        onProgress: function (bytesUploaded, bytesTotal) {
          let percentage = Math.ceil(bytesUploaded / bytesTotal * 100);

          console.log(bytesUploaded, bytesTotal, percentage + "%")
        },
        onSuccess: async () => {
          console.log('Upload success');
        }
      });

      upload.start();
```

Im a doing this wrong?

Because when using "file" instead of "file.stream()" works fine. The reason i dont wanna use "file" is because im gonna make a custom ReadableStream that modifies the data and upload it as a stream.

**Comment by Acconut** [comment]
Thank you for this report. Does the memory consumption also grow in Firefox or browsers other than Chrome?

**Comment by Acconut** [comment]
Thanks, I will look into this when I have the time.

**Comment by ISnowFoxI** [comment]
Hello, any update on this?

**Comment by Acconut** [comment]
I tested a similar setup as @Guling85 described:

```
new Upload(file.stream().getReader(), {
        chunkSize: 6553600,
        uploadLengthDeferred: true,
```

However, I did not find any signs of a memory leak or excessive memory usage in general when uploading a 1GB file. In Chrome, the heap size was around 17MB, while Firefox only reported a 3MB heap (which seems a bit odd given that we configured a buffer of `6553600` bytes).

![image](https://github.com/tus/tus-js-client/assets/1375043/cab64e3c-db87-4bfc-8deb-05db426fee2c)

![image](https://github.com/tus/tus-js-client/assets/1375043/6cce8a49-eeaa-41c9-a8af-492e452fb0b7)

@ISnowFoxI Are you experiencing memory issues with tus-js-client in the browser? If not, I will close this issue. Since it was originally opened, we also reworked parts of the client for handling stream resource. So it is possible that there was an issue in older versions.
